### PR TITLE
Horrible workaround for inference of NamedTuple constructor

### LINKED ISSUE #34: innerjoin gives odd result with missing values
I just started trying out the current incarnation of `TypedTables` for real work, I love it. The tutorial is spot on btw.

I ran into the following oddity, if `missing` is added to the `innerjoin` example:

```julia
julia> customers = Table(id = 1:3, name = ["Alice", "Bob", "Charlie"], address = ["12 Beach Street", "163 Moon Road", "6 George Street"])

julia> orders = Table(customer_id = [2, 2, 3, 3], items = ["Socks", "Tie", "Shirt", missing])

julia> innerjoin(getproperty(:id), getproperty(:customer_id), customers, orders)
4-element Array{Any,1}:
 NamedTuple{(:id, :name, :address, :customer_id, :items),Tuple{Int64,String,String,Int64,Union{Missing, String}}}((2, "Bob", "163 Moon Road", 2, "Socks"))      
 NamedTuple{(:id, :name, :address, :customer_id, :items),Tuple{Int64,String,String,Int64,Union{Missing, String}}}((2, "Bob", "163 Moon Road", 2, "Tie"))        
 NamedTuple{(:id, :name, :address, :customer_id, :items),Tuple{Int64,String,String,Int64,Union{Missing, String}}}((3, "Charlie", "6 George Street", 3, "Shirt"))
 NamedTuple{(:id, :name, :address, :customer_id, :items),Tuple{Int64,String,String,Int64,Union{Missing, String}}}((3, "Charlie", "6 George Street", 3, missing))
```

No longer a table!  What's more, this array of `Any` is not directly convertible to one either. I did the following nasty trick for now.

```julia
julia> Table(collect(NamedTuple, innerjoin(getproperty(:id), getproperty(:customer_id), customers, orders)))
Table with 5 columns and 4 rows:
     id  name     address          customer_id  items
   ┌───────────────────────────────────────────────────
 1 │ 2   Bob      163 Moon Road    2            Socks
 2 │ 2   Bob      163 Moon Road    2            Tie
 3 │ 3   Charlie  6 George Street  3            Shirt
 4 │ 3   Charlie  6 George Street  3            missing
```

### PR DESCRIPTION
**Pull request by andyferris**
Mostly fixes #34. I think I can live with waiting for a proper fix of JuliaLang/julia#29970 for more than 32 columns.

cc @c42f

### CODE REVIEW

**Comment by andyferris** [comment]
From #34:

```julia
julia> innerjoin(getproperty(:id), getproperty(:customer_id), customers, orders)
Table with 5 columns and 4 rows:
     id  name     address          customer_id  items
   ┌───────────────────────────────────────────────────
 1 │ 2   Bob      163 Moon Road    2            Socks
 2 │ 2   Bob      163 Moon Road    2            Tie
 3 │ 3   Charlie  6 George Street  3            Shirt
 4 │ 3   Charlie  6 George Street  3            missing
```

File: src/TypedTables.jl [hunk 1]
```diff
@@ -32,6 +32,18 @@ end
...
 _ndims(::Type{<:Tuple{Vararg{AbstractArray{<:Any, n}}}}) where {n} = n
 
+# Workaround for JuliaLang/julia#29970
+let
+    for n in 1:32
+        Ts = [Symbol("T$i") for i in 1:n]
+        xs = [:(x[$i]) for i in 1:n]
+        NT = :(Core.NamedTuple{names, Tuple{$(Ts...)}})
+        eval(quote
+            $NT(x::Tuple{$(Ts...)}) where {names, $(Ts...)} = $(Expr(:new, NT, xs...))
```
Comment (c42f) [inline src/TypedTables.jl]: Or equivalently
```julia
@eval $NT(x::Tuple{$(Ts...)}) where {names, $(Ts...)} = $(Expr(:new, NT, xs...))
```

**Comment by andyferris** [comment]
Seemed to fix inference but AFAICT not performance. I'm still digging in...

**Comment by andyferris** [comment]
Man, this is a real rabit hole... inference can (sometimes?) handle the `new` but I don't think `new` is good at making efficient code with union types or something.

**Comment by c42f** [comment]
Thanks Andy :+1:

**Comment by andyferris** [comment]
Don't thank me until you double-check performance doesn't fall off a cliff :)

**Comment by c42f** [comment]
Amusingly, I don't care about performance at all right now! The tables I'm using are clinical data and the number of patients is tiny...

**Comment by andyferris** [comment]
Yes.

It's an interesting realm. Typical problems with relational data are either small or so I/O bound that python seems fast. Pointclouds-as-tables are perhaps an unusual realm, but *seriously*, why shouldn't you're GIS database be as fast as hand-written C++? :)

**Comment by c42f** [comment]
I reckon it's just that languages generally don't *allow* you to have pointclouds-as-tables with any kind of usable speed or syntax. So people don't even consider that as an option and go straight for the array-of-handcoded-structs.

I think there's a lot of places this could be really cool and useful. Eg it would be great for geometry APIs for things more complicated than point clouds. TypedTables on the GPU? heck yes.

**Comment by andyferris** [comment]
Good points. And there's already no reason columns can't be GPU arrays :) I'd love to put together a demo in the "Advanced topics" docs.

**Comment by c42f** [comment]
I just tried naively throwing in some `CuArrays` and it seems there are some practical limitations.

```
julia> t = Table(a=cu(rand(10000)), b=cu(rand(10000)))
Table with 2 columns and 10000 rows:
      a            b
    ┌────────────────────────
 1  │ 0.80206746   0.3335804
 2  │ 0.1742395    0.5159153
...

julia> filter(r->r.a > 0.5, t)
ERROR: Logical indexing not implemented
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] to_index(::CuArray{Float32,1}, ::Base.LogicalIndex{Int64,Array{Bool,1}}) at /home/chris/.julia/packages/GPUArrays/hzyWn/src/indexing.jl:67
...
```

**Comment by andyferris** [comment]
Interesting - thanks.

**Pull request merged**